crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -332,56 +333,210 @@ fn build_new_replicated_loglet_configuration(
     preferred_sequencer: Option<NodeId>,
 ) -> Option<ReplicatedLogletParams> {
     let mut rng = thread_rng();
-    // todo make min nodeset size configurable, respect roles, respect StorageState, etc.
-    let nodes_config = metadata().nodes_config_ref();
+    // todo: make the default nodeset size configurable
+    let replication = previous_configuration
+        .map(|config| config.replication.clone())
+        .unwrap_or_else(|| ReplicationProperty::new(NonZeroU8::new(2).expect("to be valid")));
 
-    let log_servers: Vec<_> = observed_cluster_state
-        .alive_nodes
-        .values()
-        .filter(|node| {
-            nodes_config
-                .find_node_by_id(**node)
-                .ok()
-                .is_some_and(|config| config.has_role(Role::LogServer))
-        })
-        .collect();
+    // todo: make nodeset selection strategy configurable
+    let strategy = NodeSetSelectionStrategy::InferredFaultToleranceStrict;
+
+    let nodes_config = metadata().nodes_config_ref();
+    let preferred_nodes = previous_configuration
+        .map(|config| NodeSet::from_single(config.sequencer.as_plain()))
+        .unwrap_or_default();
 
-    let sequencer = preferred_sequencer
+    let &sequencer = preferred_sequencer
         .and_then(|node_id| {
             // map to a known alive node
             observed_cluster_state.alive_nodes.get(&node_id.id())
         })
         .or_else(|| {
             // we can place the sequencer on any alive node
             observed_cluster_state.alive_nodes.values().choose(&mut rng)
-        });
-
-    if log_servers.len() >= 3 {
-        let replication = ReplicationProperty::new(NonZeroU8::new(2).expect("to be valid"));
-        let mut nodeset = NodeSet::empty();
+        })?;
 
-        for node in &log_servers {
-            nodeset.insert(node.as_plain());
-        }
+    let selection = NodeSetSelector::new(nodes_config.as_ref(), observed_cluster_state).select(
+        strategy,
+        &replication,
+        &mut rng,
+        &preferred_nodes,
+    );
 
-        Some(ReplicatedLogletParams {
+    match selection {
+        Ok(nodeset) => Some(ReplicatedLogletParams {
             loglet_id,
-            sequencer: *sequencer.expect("one node must exist"),
+            sequencer,
             replication,
             nodeset,
             write_set: None,
-        })
-    } else if let Some(sequencer) = sequencer {
-        previous_configuration.cloned().map(|mut configuration| {
-            configuration.loglet_id = loglet_id;
-            configuration.sequencer = *sequencer;
-            configuration
-        })
```

Comment:
What happens if we cannot find a valid new nodeset (e.g. we had 3 nodes, now one is dead)? Shouldn't we be able to reconfigure using the old configuration if we were able to seal? That way we could move the sequencer to another node re-establishing write-availability.

Reply:
I totally missed that I had destroyed the handling of that case! I've restored it now; I am working on test cases for `LogsController` to cover this and other scenarios but it's a bit ugly, so I'm pushing an update containing the fix, and will follow up with the tests separately.

Reply:
I didn't get to do this on Friday, but it's back now! Does the comment below correctly capture why we do it this way?